Replace print with serialize in Immutable plugins

pedrottimark opened the pull request:
**Summary**

@cpenarrieta your feedback is welcome :) for more info, see #4114 and #4184

Breaking change in Map, OrderedMap, and Record to **print the key**:
* because implicit conversion to string throws an error if key is **symbol**
* because implicit conversion to string is `[object Object]` if key is **object**
* to distinguish **primitive values** `false` (and so on) from `"false"` as string

Breaking changes for **edge cases**:
* support `maxDepth` option
* support `indentation` of immutable collection as child of non-immutable (the closing punctuation was incorrect in the same way as for closing tag in array of React elements)

**Question**: Although I see a value in the line break between brackets/braces for non-minified empty Immutable collections (see #4121) now that `lib/immutable.js` calls the core helper functions, and the next version of Jest will have breaking changes for snapshots, it is a good time to review whether that difference is worth the extra code logic.

**Test plan**

Updated: string keys in double quote marks

Added:
* supports non-string keys
* indentation of heterogeneous collections
* indent option
* maxDepth option

pedrottimark commented:
Updated 3 snapshot tests in packages/jest-matchers/src/__tests__/spy_matchers.test.js

Replaced copy-and-paste-reuse with my first iterator with help from *Understanding ECMAScript 6*

pedrottimark commented:
AppVeyor build failed: spy_matchers.test.js 3 received values don’t show the changes to plugin

Related to #4135 ?

pedrottimark commented:
Two questions about possible additional (or should I say, subtractional) simplifications:

* Do test expressions require `!!` explicit coercion to boolean?
* Could we merge `OrderedMap` into `Map` and `OrderedSet` into `Set` plugin as follows:

```js
export const test = (val: any) => val && val[IS_MAP];
export const serialize = (
  val: any,
  config: Config,
  indentation: string,
  depth: number,
  refs: Refs,
  printer: Printer,
) =>
  printImmutableEntries(
    val,
    config,
    indentation,
    depth,
    refs,
    printer,
    val[IS_ORDERED] ? 'OrderedMap' : 'Map',
  );
```

```js
export const test = (val: any) => val && val[IS_SET];
export const serialize = (
  val: any,
  config: Config,
  indentation: string,
  depth: number,
  refs: Refs,
  printer: Printer,
) =>
  printImmutableValues(
    val,
    config,
    indentation,
    depth,
    refs,
    printer,
    val[IS_ORDERED] ? 'OrderedSet' : 'Set',
  );
```

pedrottimark commented:
And then there was one!

I worried that it’s a wrecking change, but `[ReactTestComponent, ReactElement, HTMLElement].concat(Immutable)` works whether Immutable is one plugin or an array of plugins

thymikee commented on packages/pretty-format/src/plugins/immutable.js:
```diff
@@ -0,0 +1,195 @@
+
+const IS_ITERABLE_SENTINEL = '@@__IMMUTABLE_ITERABLE__@@';
+const IS_RECORD_SENTINEL = '@@__IMMUTABLE_RECORD__@@'; // v4 or later
+
```
I think we can remove these newlines between consts and between imports

thymikee commented on packages/pretty-format/src/plugins/immutable.js:
```diff
@@ -0,0 +1,195 @@
+  printer: Printer,
+): string => {
+  if (val[IS_SEQ_SENTINEL]) {
+    return '[' + getName('Seq') + ']';
```
Do we plan to support `Immutable.Seq`?

pedrottimark replied:
Because `Record` doesn’t have a sentinel in immutable v3, `serialize` needs to account for all types that do. The unified `test` matches `Seq` too, which seems easiest to understand. Do you mean support it better?

thymikee replied:
Yup, was thinking about adding it later, if we need it. But not as a part of this PR, let's keep it tight

thymikee commented on packages/jest-matchers/src/__tests__/__snapshots__/spy_matchers.test.js.snap:
```diff
@@ -20,7 +20,7 @@ exports[`lastCalledWith works with Immutable.js objects 1`] = `
 
 Expected mock function to not have been last called with:
-  <green>[Immutable.Map {a: {\\"b\\": \\"c\\"}}, Immutable.Map {a: {\\"b\\": \\"c\\"}}]</>"
+  <green>[Immutable.Map {\\"a\\": {\\"b\\": \\"c\\"}}, Immutable.Map {\\"a\\": {\\"b\\": \\"c\\"}}]</>"
```
This is breaking, but we're more consistent 👍

thymikee commented on packages/pretty-format/src/__tests__/immutable.test.js:
```diff
@@ -438,56 +435,104 @@ describe('Immutable.OrderedMap plugin', () => {
+      '  Immutable.Map {',
+      '    "key": "value",',
+      '  }: "immutable map",',
+      '}',
```
I think at some point we should switch such tests to template literals plus [`dedent`](https://www.npmjs.com/package/dedent)

pedrottimark commented:
* Renamed `getName` as `getImmutableName`
* Moved `Seq` after all other types except `Record` (to prioritize `Map` and `List` types)
* Deleted empty lines as requested

pedrottimark commented:
Thank you for investing the time to review. I don’t take it for granted.

thymikee commented:
Still need to wait until #4135 is resolved (same thing happening in my snapshot breaking PR)

thymikee commented on packages/pretty-format/src/plugins/immutable.js:
```diff
@@ -0,0 +1,191 @@
+const getImmutableName = name => 'Immutable.' + name;
+const SPACE = ' ';
+
+const printImmutableEntries = (
```
There's a lot of duplication in `printImmutableEntries` and `printImmutableValues`, but I think that extracting common logic would make it harder to understand what's going on. Not sure but I have no strong feelings about this though. Keep it as you like
